SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -1849,52 +1881,100 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
 
-			if (GSenderStack.Num() == 0)
+			if (bIsUnordered)
 			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+				SenderInfo.Entity = WorkerEntityId;
 			}
-			else
+			else if (!bHasSenderAvailable)
 			{
-				SenderActor = GSenderStack.Last();
-			}
+				// Migration branch
+				if (bIsNetWriteFence)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error,
+						   TEXT("Net write fence will be dropped because no sender was provided. Function : %s, Target : %s"),
+						   *Function->GetName(), *Actor->GetName());
+					return;
+				}
+				else
+				{
+					UE_LOG(LogSpatialOSNetDriver, Warning,
+						   TEXT("Ordered reliable RPC will be sent unordered because no sender was provided. Function : %s, Target : %s"),
+						   *Function->GetName(), *Actor->GetName());
 
-			if (SenderActor == nullptr)
-			{
-				// These will eventually become error cases
-				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor set to call RPC %s on target %s"), *Function->GetName(),
-				//	   *Actor->GetName());
-				// return;
+					SenderInfo.Entity = WorkerEntityId;
+				}
 			}
-
-			if (SenderActor && !SenderActor->HasAuthority())
+			else
 			{
-				UE_LOG(LogSpatialOSNetDriver, Error, TEXT("No authority on sender Actor %s to call RPC %s on target %s"),
-					   *SenderActor->GetName(), *Function->GetName(), *Actor->GetName());
-				return;
-			}
+				// Long term branch
+				AActor* SenderActor = nullptr;
+
+				if (!GSenderActor.IsSet())
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor. Function : %s, Target : %s"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+				else
+				{
+					SenderActorDesc& Desc = GSenderActor.GetValue();
+					SenderActor = Desc.Actor;
+
+					if (bIsOnlyNetWriteFence && Desc.Kind != SenderActorDesc::Dependent
+						|| (!bIsNetWriteFence && Desc.Kind == SenderActorDesc::Dependent))
+					{
+						UE_LOG(LogSpatialOSNetDriver, Error,
+							   TEXT("Wrong kind of sender Actor. Check that the right AActor function was used with the right kind of RPC "
+									"(CrossServer and NetWriteFence). Function : %s, Target : %s"),
+							   *Function->GetName(), *Actor->GetName());
+						return;
+					}
+				}
+
+				GSenderActor.Reset();
+
+				if (SenderActor == nullptr)
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor. Function : %s, Target : %s"), *Function->GetName(),
+						   *Actor->GetName());
+					return;
+				}
+
+				if (!SenderActor->HasAuthority())
+				{
+					UE_LOG(LogSpatialOSNetDriver, Error, TEXT("No authority on sender Actor. Function : %s, Target : %s"),
+						   *SenderActor->GetName(), *Function->GetName(), *Actor->GetName());
+					return;
+				}
+
+				if (bIsNetWriteFence)
+				{
+					SenderActor->ForceNetUpdate();
```

Comment:
Is there an edge case here when using the replication graph where if the sending actor isn't considered relevant, than this net update won't be processed immediately anyway?

Reply:
So yeah, this is more of a declaration of intention than a proper guarantee right now.
1) We do not have component set updates
2) The way we should tackle that is by flushing a NetWriteFence queue when replicating the actor, which would allow us to bundle updates together (if that is the way the component set update API would work ?).